trie-db/src/node.rs

```diff
@@ -41,88 +42,112 @@ pub enum Node<'a> {
-		if self.has_value {
-			Some(&self.data[self.ubounds[16]..self.ubounds[17]])
-		} else {
-			None
-		}
-	}
-
-	/// Test if the node has a value.
-	pub fn has_value(&self) -> bool {
-		self.has_value
-	}
-
-	pub fn index(&self, index: usize) -> Option<&[u8]> {
-		assert!(index < 16);
-		if self.ubounds[index] == self.ubounds[index + 1] {
-			None
-		} else {
-			Some(&self.data[self.ubounds[index]..self.ubounds[index + 1]])
-		}
+impl NodePlan {
+	/// Build a node by decoding a byte slice according to the node plan. It is the responsibility
+	/// of the caller to ensure that the node plan was created for the argument data.
+	///
+	/// This may return None if the byte slice is not large enough.
+	pub fn build_node(self, data: &[u8]) -> Option<Node> {
+		let node = match self {
+			NodePlan::Empty => Node::Empty,
+			NodePlan::Leaf { partial, partial_padding, value } => {
+				let partial_slice = NibbleSlice::new_offset(data.get(partial)?, partial_padding);
+				Node::Leaf(partial_slice, data.get(value.clone())?)
+			},
+			NodePlan::Extension { partial, partial_padding, child_data } => {
+				let partial_slice = NibbleSlice::new_offset(data.get(partial)?, partial_padding);
+				Node::Extension(partial_slice, data.get(child_data.clone())?)
+			},
+			NodePlan::Branch { value, children } => {
+				let mut child_slices = [None; nibble_ops::NIBBLE_LENGTH];
+				for i in 0..nibble_ops::NIBBLE_LENGTH {
+					if let Some(range) = children[i].clone() {
+						child_slices[i] = Some(data.get(range)?);
+					}
+				}
+				let value_slice = match value {
+					Some(range) => Some(data.get(range)?),
+					None => None,
+				};
+				Node::Branch(child_slices, value_slice)
+			},
+			NodePlan::NibbledBranch { partial, partial_padding, value, children } => {
+				let partial_slice = NibbleSlice::new_offset(data.get(partial)?, partial_padding);
+				let mut child_slices = [None; nibble_ops::NIBBLE_LENGTH];
+				for i in 0..nibble_ops::NIBBLE_LENGTH {
+					if let Some(range) = children[i].clone() {
+						child_slices[i] = Some(data.get(range)?);
+					}
+				}
+				let value_slice = match value {
+					Some(range) => Some(data.get(range)?),
+					None => None,
+				};
+				Node::NibbledBranch(partial_slice, child_slices, value_slice)
+			},
+		};
+		Some(node)
 	}
 }
 
-/// An owning node type. Useful for trie iterators.
+/// An `OwnedNode` is an owned type from which a `Node` can be constructed which borrows data from
+/// the `OwnedNode`. This is useful for trie iterators.
 #[cfg_attr(feature = "std", derive(Debug))]
 #[derive(PartialEq, Eq)]
-pub enum OwnedNode {
-	/// Empty trie node.
-	Empty,
-	/// Leaf node: partial key and value.
-	Leaf(NibbleVec, DBValue),
-	/// Extension node: partial key and child node.
-	Extension(NibbleVec, DBValue),
-	/// Branch node: children and an optional value.
-	Branch(Branch),
-	/// Branch node: children and an optional value.
-	NibbledBranch(NibbleVec, Branch),
+pub struct OwnedNode<D: AsRef<[u8]>> {
```

Comment:
- 'OwnedNode' naming is confusing to me as it can now be 'OwnedNode<&'a[u8]>'.

Reply:
Changed from `D: AsRef<[u8]>` to `D: Borrow<[u8]>`.